### PR TITLE
fix(model_checkpoint): don't delete remote checkpoints outside `dirpath`

### LINKED ISSUE #21813: `ModelCheckpoint` deletes *previous run's* checkpoint when remote filesystem
### Bug description

`ModelCheckpoint` deletes the *previous run's* checkpoint, including the exact file the trainer resumed from, when the checkpoint dirpath is on a remote  (fsspec) filesystem. Authored with the help of an agent but I detected the bug myself.

I continued a training using `Trainer.fit` passing in `ckpt_path`. That run crashed and noticed that the checkpoint I resumed the training from was no longer present in the bucket (R2/S3). It was a `topk` checkpoint. Looking at the code https://github.com/Lightning-AI/pytorch-lightning/blob/4819088b0c6838f0d878b30050d59003305924b8/src/lightning/pytorch/callbacks/model_checkpoint.py#L1007-L1014

it seems intentional since it implies the protection is only for local paths. This PR https://github.com/Lightning-AI/pytorch-lightning/pull/19023 claims to have fixed it but not sure why the guard leaves out non-local file systems. Is it intentional to delete remote, pre-existing checkpoints? 
The resuming run subsequently crashed and I cant start it again since the original checkpoint is gone 😬 

 Sequence:

  1. Run 1 trains with `ModelCheckpoint(dirpath="s3://bucket/run1", every_n_train_steps=N, save_top_k=1, monitor=None)` and dies, leaving
  `run1/periodic-step=X.ckpt`.
  2. Run 2 resumes: `trainer.fit(..., ckpt_path="s3://bucket/run1/periodic-step=X.ckpt")` with a **new** dirpath `s3://bucket/run2`.
  3. At run 2's first periodic save, `run1/periodic-step=X.ckpt` is silently deleted. If it was the only copy (`save_top_k=1`), the previous run's data is  gone.

 Cause chain:

  - `ModelCheckpoint.load_state_dict` restores `best_model_path` **unconditionally**, even when it warns that the dirpath changed and skips the other fields.
  - For `monitor=None`, `_save_none_monitor_checkpoint` uses that restored `best_model_path` as `previous` and consults `_should_remove_checkpoint`.
  - `_should_remove_checkpoint` returns `True` for any non-local `previous` (`if not _is_local_file_protocol(previous): return True`) *before* the two safety   guards. Per its own docstring, "not in the current checkpoint directory" and "the checkpoint the Trainer resumed from" only protect when "the filesystem is local".

##### Expected behavior

Same as local: a checkpoint outside the callback's own `dirpath` is never deleted.

### What version are you seeing the problem on?

v2.6

### How to reproduce the bug

Self-contained — `memory://` is a non-local protocol, so no cloud credentials are needed:

<details>

```python
  import fsspec, torch
  from torch.utils.data import DataLoader, TensorDataset
  import lightning as L
  from lightning.pytorch.callbacks import ModelCheckpoint

  class Dummy(L.LightningModule):
      def __init__(self):
          super().__init__()
          self.layer = torch.nn.Linear(4, 1)
      def training_step(self, batch, _):
          x, y = batch
          return torch.nn.functional.mse_loss(self.layer(x), y)
      def configure_optimizers(self):
          return torch.optim.SGD(self.parameters(), lr=0.01)

  def loader():
      return DataLoader(TensorDataset(torch.randn(512, 4), torch.randn(512, 1)), batch_size=8)

  def cb(dirpath):
      return ModelCheckpoint(dirpath=dirpath, filename="periodic-step={step}",
                             every_n_train_steps=5, save_top_k=1,
                             auto_insert_metric_name=False, save_on_train_epoch_end=False)

  def trainer(c, steps):
      return L.Trainer(max_steps=steps, callbacks=[c], logger=False, enable_progress_bar=False,
                       enable_model_summary=False, accelerator="cpu")

  c1 = cb("memory://ckpts/run1")
  trainer(c1, 12).fit(Dummy(), loader())        # leaves run1/periodic-step=10.ckpt

  c2 = cb("memory://ckpts/run2")                # resume into a DIFFERENT dirpath
  trainer(c2, 20).fit(Dummy(), loader(), ckpt_path=c1.best_model_path)

  fs = fsspec.filesystem("memory")
  try:
      print("run1:", fs.ls("/ckpts/run1", detail=False))
  except FileNotFoundError:
      print("run1: []")
  print("run2:", fs.ls("/ckpts/run2", detail=False))
 ```

</details>

 Output:

 ```
  run1: []                                      # resumed-from checkpoint was deleted
  run2: ['/ckpts/run2/periodic-step=20.ckpt']
  ```

Replace the two `memory://` dirs with local paths and `run1/periodic-step=10.ckpt` survives, as the docstring promises.

```
run1: ['periodic-step=10.ckpt']
run2: ['periodic-step=20.ckpt']
```
run 1's checkpoint survives.

### Environment

<details>
  <summary>Current environment</summary>

- lightning: 2.6.0
- pytorch: 2.10.0
- fsspec: 2025.12.0
- Python 3.10, Linux

</details>

### More info

[Suggested fix by agent]

The "different directory" guard does not need path resolution for remote URIs — both strings are absolute URIs. Replacing the early return with a normalized containment check fixes it while preserving `save_top_k` retention inside the run's own dir:

```python
  if not _is_local_file_protocol(previous):
      return previous.startswith(self.dirpath.rstrip("/") + "/")
```

This errs toward keeping (worst case: a stale checkpoint accumulates) rather than deleting.

cc @ethanwharris

### PR DESCRIPTION
**Pull request by deependujha**
## What does this PR do?

Fixes #21813

## What does this PR do?

`ModelCheckpoint._should_remove_checkpoint` has two safety guards: never delete a checkpoint that lives outside the callback's own `dirpath`, and never delete the checkpoint the `Trainer` resumed from. Both were skipped for non-local paths:

```python
if not _is_local_file_protocol(previous):
    return True
```

So when `dirpath` is on a remote filesystem (`s3://`, `gs://`, `memory://`, ...), resuming into a *new* `dirpath` silently deleted the previous run's checkpoint at the first save, the very file the run resumed from. With `save_top_k=1` and no other copy, that data is gone.

This PR applies both guards to **`remote paths`** too, comparing them as strings rather than through `Path`:

Note: #19023 proposes a fix for the containment guard only; this PR also restores the resumed-from guard for remote paths and covers both with tests.

## Reproduction

The issue's `memory://` repro now leaves run1 intact:

```
run1: ['/ckpts/run1/periodic-step=10.ckpt']
run2: ['/ckpts/run2/periodic-step=20.ckpt']
```

Also verified end-to-end against a real S3 bucket with two runs resuming across directories.

<details>

<summary>code to reproduce & verify </summary>

```python
import fsspec
import torch
from datetime import datetime
from torch.utils.data import DataLoader, TensorDataset
import lightning as L
from lightning.pytorch.callbacks import ModelCheckpoint

timestamp = datetime.now().isoformat()
print(timestamp)

class Dummy(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(4, 1)

    def training_step(self, batch, _):
        x, y = batch
        return torch.nn.functional.mse_loss(self.layer(x), y)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)


def loader():
    return DataLoader(
        TensorDataset(torch.randn(512, 4), torch.randn(512, 1)), batch_size=8
    )


def cb(dirpath):
    return ModelCheckpoint(
        dirpath=dirpath,
        filename="periodic-step={step}",
        every_n_train_steps=5,
        save_top_k=1,
        auto_insert_metric_name=False,
        save_on_train_epoch_end=False,
    )


def trainer(c, steps):
    return L.Trainer(
        max_steps=steps,
        callbacks=[c],
        logger=False,
        enable_progress_bar=False,
        enable_model_summary=False,
        accelerator="cpu",
    )


path1 = f"s3://lightning-checkpoints/pl-model-checkpoint-21813/{timestamp}/run1"
path2 = f"s3://lightning-checkpoints/pl-model-checkpoint-21813/{timestamp}/run2"
c1 = cb(path1)
trainer(c1, 12).fit(Dummy(), loader())  # leaves run1/periodic-step=10.ckpt
print("-"*100)
c2 = cb(path2)  # resume into a DIFFERENT dirpath
trainer(c2, 20).fit(Dummy(), loader(), ckpt_path=c1.best_model_path)

fs = fsspec.filesystem("s3")
try:
    print("run1:", fs.ls(path1, detail=False))
except FileNotFoundError:
    print("run1: []")
print("run2:", fs.ls(path2, detail=False))
```
</details>

- expected output
```
run1: ['lightning-checkpoints/pl-model-checkpoint-21813/2026-07-30T12:04:24.525938/run1/periodic-step=10.ckpt']
run2: ['lightning-checkpoints/pl-model-checkpoint-21813/2026-07-30T12:04:24.525938/run2/periodic-step=20.ckpt']
```

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/CHANGELOG.md [hunk 1]
```diff
@@ -32,6 +32,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed `ModelCheckpoint` deleting checkpoints outside its own `dirpath`, including the one the `Trainer` resumed from, when the checkpoints live on a remote filesystem ([#21813](https://github.com/Lightning-AI/pytorch-lightning/issues/21813))
```
Comment (deependujha) [inline src/lightning/pytorch/CHANGELOG.md]:
```suggestion
- Fixed `ModelCheckpoint` deleting checkpoints outside its own `dirpath`, including the one the `Trainer` resumed from, when the checkpoints live on a remote filesystem ([#21867](https://github.com/Lightning-AI/pytorch-lightning/issues/21867))
```

**Comment by deependujha** [comment]
**md linkcheck issue**:

https://huggingface.co/datasets/cerebras/SlimPajama-627B is no longer available.

https://huggingface.co/datasets/gmongaras/SlimPajama-627B_Reupload is a reuploaded version, but not sure if it's safe to replace it with.

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -1009,19 +1009,24 @@ def _should_remove_checkpoint(self, trainer: "pl.Trainer", previous: str, curren
...
+        - The previous checkpoint is the checkpoint the Trainer resumed from
 
         """
         if previous == current:
             return False
+        assert self.dirpath is not None
         if not _is_local_file_protocol(previous):
-            return True
+            previous_uri = previous.replace("\\", "/")
+            if trainer.ckpt_path is not None and previous_uri == str(trainer.ckpt_path).replace("\\", "/"):
```
Comment (Sanjays2402) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:1021]: the remote branch compares `trainer.ckpt_path` by exact string, but when the resume path came from `ckpt_path="last"` it went through `_find_last_checkpoints`, which returns `os.path.normpath(p)` over `self._fs.ls(...)`. on memory/s3 that gives `/ckpts/run1/last.ckpt` (no scheme), and even if a scheme survives normpath collapses `memory://` to `memory:/`. so `previous_uri == str(trainer.ckpt_path)` is false and the resumed-from checkpoint gets deleted anyway. the local branch is immune because both sides go through `Path(...).absolute()`. worth normalizing both with `url_to_fs`/`strip_protocol` before comparing?
